docs/internal/generics.md

```diff
@@ -0,0 +1,114 @@
+# Generics
+
+Generics are a way to use type parameters to define or restrict other types.
+For example, a generic class can use its type parameters to define the type of fields, and a generic method can use its parameters to restrict what can be passed as an argument.
+
+Example:
+```ts
+class MyGenericType[MyType, SomeInterface]
+{
+  // MyType is the type of this field
+  field : MyType;
+
+  // SomeInterface is the super type of anything that can be passed to this method
+  method(arg : SomeInterface)
+}
+
+// Some concrete types that implements Hashable
+class MyClass : Hashable { ... }
+class MyOtherClass : Hashable { ... }
+
+// This creates a concrete type
+let x : MyGenericType[U32, Hashable] & mut = MyGenericType;
+
+x.field = 42; // OK: The type of the field is compatible with integral literals
+
+let c = MyClass;
+x.method(c); // OK: MyClass implements Hashable
+let o = MyOtherClass;
+x.method(o); // OK: MyOtherClass also implements Hashable
+
+// This creates another concrete type, now with a more restrictive argument
+let y : MyGenericType[F64, MyClass] = MyGenericType;
+
+y.field = 42; // ERROR: No known conversion between integer literals and float types
+
+y.method(c); // OK: This is the expected type
+y.method(o); // ERROR: Even though MyOtherType implements the same interface, it's not he same type
+```
+
+In the example above, there is one generic class called `MyGenericClass` and two implementations (or specialisations) with types `[U32, Hashable]` and `[F64, MyClass]`.
+If all types are used correctly (ex. removing the ERROR lines above), the compiler will create two concrete classes from the one generic class: one for each specialisation.
+
+## Type Parameters
+
+Type parameters are the named types that can be used inside a generic declaration in place of actual types.
```

Comment:
Do we allow any kind of value parameters to generics?  

The simplest use for simple value-dependent types is to specify fixed-size arrays.  For example, `Array[U32, 10]` takes on generic parameter that is a type and one that is a value.  

C++ has always allowed templates to take primitive value types as parameters and with C++20 allows class literals to be template parameters (with the type defined by the type of the class and the byte pattern that represents a specific instance).  I don't think we'd want to take the same route as C++20 because it composes in very exciting ways with out desire to have arbitrary compile-time evaluation.  

You can trivially encode integers as church numerals to map from the value space to the type space, though it's incredibly clunky.  For example:

```verona
class Zero {}
class N[T] {}

// If programmers ever have to write this we have done something wrong!
var arr : Array[U32, N[N[N[N[N[N[N[N[N[N[Zero]]]]]]]]]]];
```

Reply:
I'm personally not against values as template parameters, but I'm not sure what types would they be and how would the type system cope with that: naturally or artificially through special cases.

Reply:
Welcome to "dependent typing", where types can bind values; it's definitely the strangest dimension of the lambda cube.  I don't have an answer, but just some background (you've given me an excuse and I enjoy thinking about these things more than I would enjoy munging C++ at the moment)...

A Haskeller's view of, say, `MyClass[5]` is that that `5` is now a _type_, unrelated to any value `5` (in particular the value `5` does not have type `5`; in fact the type `5` has no values); one can introduce a notion of "kinds" of types and think of that type `5` as having _kind_ `int` and the _type_ constructor `MyClass` as having kind `int -> *`, where `*` is the kind of types that have values (there's also the _value_ constructor `MyClass` that takes values and builds a `MyClass`-typed value).  Because the type-level ints and value-level ints are different objects, while we can have a polymorphic "unlift" value which is the value `n` when instantiated at type `n`, its type (`forall _ : int . int`) would also be inhabited by, say, a constant `0` value, which might be dissatisfying.  On the other hand, because we've bolted type-level ints on to the side, the flavor of the type system is not hugely changed, and, notably, does not necessarily require a subtyping system, which may keep type elaboration/inference (and checking) tractable.

If we want to keep the notion of types as sets of values, we could employ singleton types, where the value `5` has type `5` (sometimes written `Sing 5` or such to make it clearer whether we mean the value or the type).  The type `U32` might then be something like an alias for the union `Sing 0 | Sing 1 | Sing 2 | ...` or might be another axiomatic type (that is, not defined algebraically, but presumably then taken to have a supertype relationship with `Sing 5`).  Value-level operators naturally lift across `Sing` and type algebra (e.g., `(Sing 0 | Sing 1) + (Sing 2) = (Sing 2 | Sing 3)`).  The analogous "unlift" polymorphic value could now be typed as `forall _ : Sing k . k`, which admits exactly the term we want and no other, which may be more satisfying.  On the other hand, this approach threatens to require a fully set-theoretic subtyping relationship, which in turn might mean that type inference is intractable.

Whether you think the "type ints aren't value ints" or "type ints are the singletons of value ints" approach is cleaner is probably mostly subjective or might depend on whether you believe that PLs are for programming computers and math is that neat thing over there or are for doing mathematics and computers are a fun afterthought -- or, at least, I'm not a sufficiently well-versed PL theorist to have a strong reason to prefer one to the other.  Both approaches run into challenges with increasing expressivity if the compiler is also required to terminate, so, probably, the question itself is undecidable; blame Turing and Gödel.

Reply:
@nwf, thanks for the digression, that was most interesting!

I'm still unsure what the benefits of value parameters have with a compiler that can already do whole program analysis and create specialisations even when the user didn't request.

For example, what's the difference between:
```ts
// Imagine this is the syntax for value parameters
class List[Ty, Len = U32] {
  // Whatever Verona equivalent to: alignas(Ty) char data[Len * sizeof(Ty)];
  data : Ty[Len];
  // Constructor
  create() {
    data = new Ty[Len]; // whatever the syntax here... (**)
  }
};
...
// Creates a list with a data section size = 5 * sizeof(MyClass)
let array = List[MyClass, 5];
```

and

```ts
class List[Ty] {
  // This is now an array of unknown size until initialisation
  data : Ty[];
  // Constructor
  create(Len : U32) {
    data = new Ty[Len]; // whatever the syntax here... (**)
  }
};
...
// Creates a list with a data section size = 5 * sizeof(MyClass)
let array = List[MyClass](5);
```

Unless the initialisation syntax (**) is different for declared size vs undeclared size, these two are the same.

If whole-program reachability is allowed to create specialisations of known patterns and it recognises that both cases are the same at compile time, we can emit the exact same code, probably inlined, even if each went through a different concrete (anonymous) specialisation type in the higher level representation.

Am I missing anything?

Reply:
It's possible that some (much?) of the _optimization_ opportunities nominally afforded by dependent typing could be realized instead by whole-program compilation, as you say.  That in and of itself is probably an interesting research topic.

But, to the example at hand, a variable whose type is `List[MyClass,5]` guarantees that it will _always_ have five elements, no matter how it is used in the source, while a variable of type `List[MyClass]` _might_ happen to always have five elements.  Whether or not the latter property could be discovered by the compiler and used for generating code, the former conveys something much more precise about programmer intent.

Reply:
Think about the simplest case: fixed-size arrays.  Imagine two possible generic types for arrays:

```verona
class Array[T] {...}
class Array[T, Size : U32] {...}
```

In the first case, size is fixed at creation time, but I have no way of specifying in a function signature that I want an array of a fixed size.  I then need to add dynamic checks in my function and the caller needs to check the docs.  The compiler can optimise away the checks if it can do sufficient dataflow propagation, but consider something like this:

```verona
var x = Array[U32](1, 2, 3);
when (whatever)
{
  whatever.foo(x);
}
```

Now it's very difficult for the compiler to infer that the size that was set in one behaviour is guaranteed in the other.  First we need to inline `Array[U32]::Create` then we need to extract the information that a `let` field has a known value, then we need to propagate that to the function implementing the `when` clause.  Reachability helps a bit here, but it's a dataflow problem not a control flow problem so reachability is not a complete solution.

Now imagine that the `foo()` method was declared as:

```
foo(arr : Array[U32, 4] & readonly) : U32;
```

Now we know at type-checking time that this is an error, because I'm passing an `Array[U32,3]` to a function that expects an `Array[U32,4]`.  Once I correct the caller to add the missing element, the compiler doesn't need to do anything to propagate the bounds: they're part of the type.

Reply:
Right, but the main use case (a big example is LLVM's `SmallVector`) is a `List` that is at least 5, with higher cost for more.

What it does is allocate statically a list of N elements up front but allow it to grow if needed.

This avoid the high costs of using `std::vector` for "small" lists with power-of-two reallocation in the case:
```c++
std::vector<int> foo;
foo.push_back(1);
foo.push_back(2);
foo.push_back(3);
foo.push_back(4);
````
Which will go from 0->1->2->4 (worst case, 3 reallocs).

In C++, even if we use `resize`, the main difference is:
* `SmallVector[Ty, N]()` -> `char data[N * sizeof(Ty)] (constexpr)`
* `std::vector[Ty]().resize(N)` -> `char* data = malloc(N * sizeof(Ty)) + free()` (even if N is constexpr)

`SmallVector` allocates more space if you go past its limit (with new/delete) and check the index to retrieve/update values on either storage.

My point is that **/maybe/** all "interesting" cases to use values as template parameters could be reached by compiler analysis (not optimisation) via constant propagation, reachability, etc. Or, put another way, the uncommon / uninteresting cases can suffer some run-time penalties to avoid having to add values to type parameters.

I'm honestly just guessing, so feel free to correct me any time. :)

Reply:
The main advantage is nothing to do with optimisation, it's to do with writing correct code:

In some cases, you want to write a function that takes a collection of *n* things, where *n* is an arbitrary number.  In other cases, you want to write a function that takes a collection of *n* things where *n* is a fixed value.

In C++, `std::vector` provides a convenient way (well, *a* way) of handling the first, `std::array` provides a convenient way of handling the second.  If I want to write a function that takes a 3x3 convolution matrix, for example, then it's *really* helpful if I can specify that I require a 9-element array in the signature of a function.  If the only tool that I have is a fixed-but-unknown-at-compile-time array, then I can't express this and I'm reliant on complex analyses to (hopefully) give me an error.

The fact that the compiler can now more easily elide bounds checks is a nice bonus but it's not the main motivation.

Reply:
Just to be clear, I wasn't advocating the performance/optimisation side only, though I understand why both of you thought so. I was just wondering if the compiler could *always* infer correctly without the need of value parameters.

I understand the difference between `List[I32, N]` and `List[I32](N)` in your examples and I agree they're fundamentally different for the type system. I'm not discussing that.

The main point I was missing is the function signature. I can easily create types that have static sizes, but I can't easily *require* types with a specific compile-time size if I can't specify its size at compile-time in a function signature.

Reply:
I've added an optional section on values as type parameters.